File: src/factories/ExpressionFactory.ts
import ts from "typescript";

import { RandomGenerator } from "../utils/RandomGenerator";

export namespace ExpressionFactory {
    export const bigint = (value: number | bigint) =>
        ts.factory.createCallExpression(
            ts.factory.createIdentifier("BigInt"),
            undefined,
            [ts.factory.createIdentifier(value.toString())],
        );

    export const isRequired = (input: ts.Expression): ts.Expression =>
        ts.factory.createStrictInequality(
            ts.factory.createIdentifier("undefined"),
            input,
        );

    export const isArray = (input: ts.Expression): ts.Expression =>
        ts.factory.createCallExpression(
            ts.factory.createIdentifier("Array.isArray"),
            undefined,
            [input],
        );

    export const isObject =
        (options: { checkNull: boolean; checkArray: boolean }) =>
        (input: ts.Expression): ts.Expression => {
            const conditions: ts.Expression[] = [
                ts.factory.createStrictEquality(
                    ts.factory.createStringLiteral("object"),
                    ts.factory.createTypeOfExpression(input),
                ),
            ];
            if (options.checkNull === true)
                conditions.push(
                    ts.factory.createStrictInequality(
                        ts.factory.createNull(),
                        input,
                    ),
                );
            if (options.checkArray === true)
                conditions.push(
                    ts.factory.createStrictEquality(
                        ts.factory.createFalse(),
                        ts.factory.createCallExpression(
                            ts.factory.createIdentifier("Array.isArray"),
                            undefined,
                            [input],
                        ),
                    ),
                );

            return conditions.length === 1
                ? conditions[0]!
                : conditions.reduce((x, y) =>
                      ts.factory.createLogicalAnd(x, y),
                  );
        };

    export const isInstanceOf =
        (type: string) =>
        (input: ts.Expression): ts.Expression => {
            return ts.factory.createBinaryExpression(
                input,
                ts.factory.createToken(ts.SyntaxKind.InstanceOfKeyword),
                ts.factory.createIdentifier(type),
            );
        };

    export const coalesce =
        (x: ts.Expression) =>
        (y: ts.Expression): ts.Expression =>
            ts.factory.createBinaryExpression(
                x,
                ts.factory.createToken(ts.SyntaxKind.QuestionQuestionToken),
                y,
            );

    export const currying =
        (target: ts.Expression) => (parameters: ts.Expression[]) => {
            if (parameters.length === 0)
                return ts.factory.createCallExpression(
                    target,
                    undefined,
                    undefined,
                );
            let prev: ts.CallExpression = ts.factory.createCallExpression(
                target,
                undefined,
                [parameters[0]!],
            );
            for (const param of parameters.slice(1))
                prev = ts.factory.createCallExpression(prev, undefined, [
                    param,
                ]);
            return prev;
        };

    export const selfCall = (body: ts.ConciseBody) =>
        ts.isCallExpression(body)
            ? body
            : ts.factory.createCallExpression(
                  ts.factory.createParenthesizedExpression(
                      ts.factory.createArrowFunction(
                          undefined,
                          undefined,
                          [],
                          undefined,
                          undefined,
                          body,
                      ),
                  ),
                  undefined,
                  undefined,
              );

    export const getEscapedText =
        (printer: ts.Printer) =>
        (input: ts.Expression): string =>
            printer.printNode(
                ts.EmitHint.Expression,
                input,
                input.getSourceFile(),
            );

    export const transpile =
        (context: ts.TransformationContext) =>
        (script: string) =>
        (input: ts.Expression): ts.Expression => {
            const file: ts.SourceFile = ts.createSourceFile(
                `${RandomGenerator.uuid()}.ts`,
                script,
                ts.ScriptTarget.ESNext,
                true,
                ts.ScriptKind.TS,
            );
            const statement: ts.Statement | undefined = file.statements[0];
            if (statement === undefined)
                throw new ReferenceError(
                    "Error on ExpressionFactory.transpile(): no statement exists.",
                );
            else if (!ts.isExpressionStatement(statement))
                throw new TypeError(
                    "Error on ExpressionFactory.transpile(): statement is not an expression statement.",
                );

            const visitor = (node: ts.Node): ts.Node => {
                if (ts.isIdentifier(node) && node.text === "$input")
                    return input;
                return ts.visitEachChild(
                    (ts.factory as any).cloneNode(node),
                    visitor,
                    context,
                );
            };
            return visitor(statement.expression) as ts.Expression;
        };
}

File: packages/core/src/factories/ExpressionFactory.go
package factories

import (
	"errors"
	"math"
	"strings"
)

type expressionFactory struct{}

var ExpressionFactory expressionFactory

func (expressionFactory) Number(value float64) Expression {
	if value < 0 {
		return Node("prefixUnary", Field("operator", "-"), Field("operand", Node("number", Field("value", math.Abs(value)))))
	}
	return Node("number", Field("value", value))
}

func (expressionFactory) Bigint(value any) Expression {
	return Node("call", Field("expression", Identifier("BigInt")), Field("arguments", []Expression{Identifier(toString(value))}))
}

func (expressionFactory) IsRequired(input Expression) Expression {
	return Node("strictInequality", Field("left", Identifier("undefined")), Field("right", input))
}

func (expressionFactory) IsArray(input Expression) Expression {
	return Node("call", Field("expression", Identifier("Array.isArray")), Field("arguments", []Expression{input}))
}

func (expressionFactory) IsObject(props IsObjectProps) Expression {
	conditions := []Expression{Node("strictEquality", Field("left", StringLiteral("object")), Field("right", Node("typeof", Field("expression", props.Input))))}
	if props.CheckNull {
		conditions = append(conditions, Node("strictInequality", Field("left", Null()), Field("right", props.Input)))
	}
	if props.CheckArray {
		conditions = append(conditions, Node("strictEquality", Field("left", False()), Field("right", ExpressionFactory.IsArray(props.Input))))
	}
	return reduceExpression("logicalAnd", conditions)
}

func (expressionFactory) IsInstanceOf(typeName string, input Expression) Expression {
	return Node("instanceOf", Field("input", input), Field("type", Identifier(typeName)))
}

func (expressionFactory) Coalesce(x Expression, y Expression) Expression {
	return Node("coalesce", Field("left", x), Field("right", y))
}

func (expressionFactory) Currying(props CurryingProps) Expression {
	if len(props.Arguments) == 0 {
		return Node("call", Field("expression", props.Function))
	}
	prev := Node("call", Field("expression", props.Function), Field("arguments", []Expression{props.Arguments[0]}))
	for _, param := range props.Arguments[1:] {
		prev = Node("call", Field("expression", prev), Field("arguments", []Expression{param}))
	}
	return prev
}

func (expressionFactory) SelfCall(body Expression, typeNode ...TypeNode) Expression {
	if body.Kind == "call" {
		return body
	}
	fields := []NodeField{Field("body", body)}
	if len(typeNode) != 0 {
		fields = append(fields, Field("type", typeNode[0]))
	}
	return Node("call", Field("expression", Node("parenthesized", Field("expression", Node("arrowFunction", fields...)))))
}

func (expressionFactory) GetEscapedText(input Expression) string {
	return input.String()
}

func (expressionFactory) Transpile(props TranspileProps) (func(Expression) Expression, error) {
	script := strings.TrimSpace(props.Script)
	if script == "" {
		return nil, errors.New("Error on ExpressionFactory.transpile(): no statement exists.")
	}
	return func(input Expression) Expression {
		return Node("transpiledExpression", Field("script", script), Field("input", input))
	}, nil
}
